[4.0] Adding a mail templating system

Hackwar opened the pull request:
This PR adds a way to modify the mails that are send by the system in a GUI and with a larger range of possibilities than with the current system of simple translation strings. Extensions have to register the mails that they want to send and then can use an API to render those mails.

### Registering a mail template
An extension has to add its templates via `MailTemplate::addTemplate()` during installation or whenever it is doing its setup. This inserts a row for each mail template into #__mail_templates, with a template_id, an empty language field, translation strings for subject, body and optionally htmlbody and in the params field the available tags in the template. The template_id has to be of the format {extension}.{sub-identifier}, for example "com_config.test_mail". This is not restricted to components, but it requires a translation file in the administrator language folder.
If there is no modified mail template, the translated translation strings are used to send the mails.

### Editing mail templates
The GUI is located in com_mails and is in the System menu in the backend below the global configuration item. All the templates are listed in that view and you can then click on the respective flag to edit the template for that language. You can enable or disable editing the different fields. The idea behind this is, that you can revert changes to the templates by going back to the original translation strings.
Next to the editors, you have the available tags, which can be inserted into the text by clicking on them. These also work in the subject, but have not been listed here separately.

### Sending a mail
To send a mail, the code could look like this:
```
$mailer = new MailTemplate('com_config.test_mail', $user->getParam('language', $app->get('language')), $mail);
$mailer->addTemplateData(array('sitename' => $app->get('sitename'), 'method' => Text::_('COM_CONFIG_SENDMAIL_METHOD_' . strtoupper($mail->Mailer))));
$mailer->addRecipient($app->get('mailfrom'), $app->get('fromname'));
$mailer->send();
```
You create a MailTemplate object with the ID of your mail template and the language that you want this in, then add the data that should be replaced in the template as associative array and the recipient and then send it.

### Features
1. Attachments
   You can attach files to your templates both statically when editing the template in the backend, as well as dynamically when sending the mail. The fields in the backend only show up when a valid path is inserted in the components configuration.
2. Plaintext & HTML mails
   Mails can be created as plaintext and HTML mails and also HTML with plaintext as fallback. Again a feature that can be selected in the components configuration.
3. Alternative mail settings
   Each template can be send via its own mailing settings, so you can set your own sending address and even your own SMTP settings.
4. Send copies of each mail
   Each mail can be send to another recipient via BCC for archiving or error checking purposes.
5. Nested templates
   A template can contain nested templates. Tags are an associative array of key-value-pairs. If the value is an array containing arrays, then the system will look for a part with `{key}...{/key}` and repeat that for every array in the array... I guess you catch what I'm trying to say. This can be nested infinitely deep.

### How to test
Right now there is one template in the system for the test mail in the global configuration. You can modify the template for your language (install additional languages and create content languages for more languages in the templates) and then test it by clicking the test button in the mail settings.

I've rewritten this PR to include features that were previously only advertised but not included and did some more cleanup work.

I'm looking forward to your feedback.

### Credits
This PR is supported by [Showcast.de](https://www.showcast.de), where we have been using a similar system for quite some time now.

Hackwar committed:
Adding a mail templating system

brianteeman commented on administrator/components/com_config/Controller/MailController.php:
```diff
@@ -0,0 +1,129 @@
+use Joomla\CMS\Language\Text;
+use Joomla\CMS\Router\Route;
+use Joomla\CMS\Session\Session;
+use Joomla\CMS\Factory;
```
please alphasort the use statements

Hackwar replied:
Done.

brianteeman commented on administrator/components/com_config/Controller/MailController.php:
```diff
@@ -0,0 +1,129 @@
+class MailController extends FormController
+{
+	/**
+	 * Method override to check if you can add a new record.
```
override?

Hackwar replied:
Done.

brianteeman commented on administrator/components/com_config/Controller/MailController.php:
```diff
@@ -0,0 +1,129 @@
+	}
+
+	/**
+	 * Method to check if you can add a new record.
```
shouldnt this be edit not add?

Hackwar replied:
Done.

brianteeman commented on administrator/components/com_config/Controller/MailController.php:
```diff
@@ -0,0 +1,129 @@
+	 *
+	 * @return  boolean
+	 *
+	 * @since   1.6
```
please update to __DEPLOY_VERSION__

Hackwar replied:
Done.

brianteeman commented on administrator/components/com_config/Model/MailModel.php:
```diff
@@ -0,0 +1,222 @@
+use Joomla\Registry\Registry;
+use Joomla\Utilities\ArrayHelper;
+use Joomla\CMS\MVC\Model\AdminModel;
+use Joomla\CMS\Object\CMSObject;
```
please alpahasort

Hackwar replied:
Done.

brianteeman commented on administrator/components/com_config/Model/MailsModel.php:
```diff
@@ -0,0 +1,119 @@
+	protected function getListQuery()
+	{
+		// Create a new query object.
+		$db		= $this->getDbo();
```
please remove extra tabs

Hackwar replied:
Done.

brianteeman commented on administrator/components/com_config/View/Mail/HtmlView.php:
```diff
@@ -0,0 +1,122 @@
+	public function display($tpl = null)
+	{
+		$this->state	= $this->get('State');
+		$this->item		= $this->get('Item');
```
please remove extra tab

Hackwar replied:
Done.

brianteeman commented on administrator/components/com_config/View/Mail/HtmlView.php:
```diff
@@ -0,0 +1,122 @@
+		$this->state	= $this->get('State');
+		$this->item		= $this->get('Item');
+		$this->master	= $this->get('Master');
+		$this->form		= $this->get('Form');
```
please remove extra tab

Hackwar replied:
Done.

brianteeman commented on administrator/components/com_config/View/Mails/HtmlView.php:
```diff
@@ -0,0 +1,123 @@
+use Joomla\CMS\Toolbar\Toolbar;
+use Joomla\CMS\Toolbar\ToolbarHelper;
+use Joomla\CMS\Language\Text;
+use Joomla\CMS\Language\Multilanguage;
```
please alphasort the use statements

Hackwar replied:
Done.

brianteeman commented on administrator/components/com_config/forms/mail.xml:
```diff
@@ -0,0 +1,84 @@
+			buttons="false"
+			class="inputbox"
+			filter="JComponentHelper::filterText"
+			label="Body"
```
Please chnage to a JTEXT key and move after type

Hackwar replied:
Done

brianteeman commented on administrator/components/com_config/forms/mail.xml:
```diff
@@ -0,0 +1,84 @@
+			buttons="false"
+			class="inputbox"
+			filter="JComponentHelper::filterText"
+			label="Body"
```
Please chnage to a JTEXT key and move after type

Hackwar replied:
Done

brianteeman commented on administrator/components/com_config/forms/mail.xml:
```diff
@@ -0,0 +1,84 @@
+		<field
+			name="attachments"
+			type="text"
+			label="Attachments"
```
Please change to a JTEXT key

brianteeman commented on administrator/components/com_config/forms/mail.xml:
```diff
@@ -0,0 +1,84 @@
+		<field
+			name="params"
+			type="text"
+			label="Params"
```
Please change to a JTEXT key

Hackwar replied:
Done

brianteeman commented on administrator/components/com_config/tmpl/mail/edit.php:
```diff
@@ -0,0 +1,99 @@
+use Joomla\CMS\Language\Text;
+use Joomla\CMS\Router\Route;
+use Joomla\CMS\Factory;
+use Joomla\CMS\HTML\HTMLHelper;
```
please alphasort

Hackwar replied:
Done.

brianteeman commented on administrator/components/com_config/tmpl/mail/edit.php:
```diff
@@ -0,0 +1,99 @@
+
+$doc = Factory::getDocument();
+$doc->addScriptDeclaration('
+jQuery(document).ready(function() {
```
this should be in vanilla es6 not jquery afaik

Hackwar replied:
Done.

brianteeman commented on administrator/components/com_config/tmpl/mails/default.php:
```diff
@@ -0,0 +1,104 @@
+use Joomla\CMS\Language\Text;
+use Joomla\CMS\Router\Route;
+use Joomla\CMS\Layout\LayoutHelper;
+use Joomla\CMS\HTML\HTMLHelper;
```
please alphasort

Hackwar replied:
Done.

brianteeman commented on administrator/components/com_config/tmpl/mails/default.php:
```diff
@@ -0,0 +1,104 @@
+JHtml::_('behavior.multiselect');
+
+$user		= JFactory::getUser();
+$userId		= $user->get('id');
```
please remove extra tab

Hackwar replied:
Done.

brianteeman commented on administrator/components/com_config/tmpl/mails/default.php:
```diff
@@ -0,0 +1,104 @@
+$userId		= $user->get('id');
+$listOrder	= $this->escape($this->state->get('list.ordering'));
+$listDirn	= $this->escape($this->state->get('list.direction'));
+$canOrder	= $user->authorise('core.edit.state', 'com_weblinks.category');
```
com_weblinks ??

Hackwar replied:
Done.

brianteeman commented on administrator/components/com_config/tmpl/mails/default.php:
```diff
@@ -0,0 +1,104 @@
+					<table class="table" id="mailList">
+						<thead>
+							<tr>
+								<th scope="col" style="min-width:100px" class="nowrap">
```
please remove nowrap - its not needed

Hackwar replied:
Done.

brianteeman commented on administrator/components/com_config/tmpl/mails/default.php:
```diff
@@ -0,0 +1,104 @@
+							list($component, $sub_id) = explode('.', $item->mail_id, 2);
+							?>
+							<tr class="row<?php echo $i % 2; ?>">
+								<td scope="row">
```
1. scope=row can only go on a th not a td
2. scope =row should only be on one cell - the one that describes the cells in the row eg title

Hackwar replied:
Done.

brianteeman commented on administrator/language/en-GB/en-GB.com_config.ini:
```diff
@@ -142,21 +142,29 @@ COM_CONFIG_FIELD_VALUE_SMTP="SMTP"
+COM_CONFIG_HEADING_LANUAGES="Languages"
 COM_CONFIG_LOCATION_SETTINGS="Location Settings"
 COM_CONFIG_MAIL_SETTINGS="Mail Settings"
+COM_CONFIG_MAIL_TEST_MAIL_DESC="This mail is send when you click the &quot;Send Test Mail&quot; button in the Global Configuration. It is sent to the sending mail address that is set in the mail settings."
```
sent not send

Hackwar replied:
Done.

brianteeman commented on libraries/src/Mail/MailTemplate.php:
```diff
@@ -0,0 +1,193 @@
+
+use Joomla\CMS\Factory;
+use Joomla\CMS\Mail\Mail;
+use Joomla\CMS\Language\Text;
```
please alphasort

Hackwar replied:
Done.

brianteeman commented on libraries/src/Mail/MailTemplate.php:
```diff
@@ -0,0 +1,193 @@
+	 */
+	public function send()
+	{
+		$db = \JFactory::getDBO();
```
please change to Factory

Hackwar replied:
Done.

Hackwar committed:
Further work on the mail templating system

Hackwar committed:
Changes from @brianteeman review
Codestyle

Hackwar committed:
Merge branch '4.0-dev' into j4mailtemplates

Hackwar committed:
Codestyle

Hackwar committed:
Merge branch 'j4mailtemplates' of github.com:Hackwar/joomla-cms into j4mailtemplates

Fedik commented:
that a cool feature!

Quy commented on administrator/components/com_config/Controller/MailController.php:
```diff
@@ -0,0 +1,302 @@
+/**
+ * The mail controller
+ *
+ * @since  DEPLOY_VERSION
```
It should be `__DEPLOY_VERSION__`

Hackwar replied:
Done

Quy commented on administrator/components/com_config/Model/MailModel.php:
```diff
@@ -0,0 +1,227 @@
+	public function getForm($data = array(), $loadData = true)
+	{
+		// Initialise variables.
+		$app	= Factory::getApplication();
```
Remove extra spaces before `=`

laoneo commented:
Instead of using a table, can we not work with plugin events to collect the available mail templates? It would make the thing more dynamic as often the strings are not static.
As components, which do have that functionality already self made, in their submenu and custom fields and categories are there too, would it be possible to offer mail template menu links for a component. The idea behind it is that people install a component and then do set it up. So it would be good when they do not have to leave the context of the component and do navigate away from it to set up the mails.

Beside that I really like that feature!

Quy commented on administrator/components/com_config/Controller/MailController.php:
```diff
@@ -0,0 +1,302 @@
+	 *
+	 * @since   DEPLOY_VERSION
+	 */
+	protected function allowAdd($data = array())
```
Use short-hand notation for array `[]`

Hackwar replied:
done

Hackwar commented:
@laoneo I thought about several ways how to do this, but storing the templates in a table is the only thing that doesn't escalate into a horrible abomination of system. If you are talking about the available tags: You can already extend those via a plugin event when sending the mails and I'm working on adding the event when rendering the tags.

Hackwar committed:
Further work on mail templates

wojsmol commented on administrator/components/com_config/Table/MailTable.php:
```diff
@@ -0,0 +1,35 @@
+/**
+ * Mail Table class.
+ *
+ * @since  4.0.0
```
`@since   __DEPLOY_VERSION__`

Hackwar replied:
Fixed

wojsmol commented on administrator/components/com_config/helpers/html/config.php:
```diff
@@ -0,0 +1,50 @@
+/**
+ * Mailtags HTML helper class.
+ *
+ * @since  4.0.0
```
`@since   __DEPLOY_VERSION__`

Hackwar replied:
Fixed

wojsmol commented on administrator/components/com_config/helpers/html/config.php:
```diff
@@ -0,0 +1,50 @@
+	 *
+	 * @return  string  List of tags that can be inserted into a field.
+	 *
+	 * @since   4.0.0
```
`@since   __DEPLOY_VERSION__`

Hackwar replied:
Fixed

Hackwar committed:
Further work on Mail Templates

Hackwar committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into j4mailtemplates

Hackwar committed:
Further work on Mail Templates

Hackwar committed:
Codestyle

Fedik commented:
When editing a template, somehow I got  a broken editor, not sure if it related to the PR.

Sending cannot test currently, have to fix my email config.

Hackwar commented:
The broken editor is a general issue when you have 2 different editors on the same page. That is something that we have to look into in another issue.

brianteeman commented:
Pretty sure that was resolved in J3

Hackwar committed:
Adding test mail as example for a mail template

Fedik commented on administrator/components/com_config/helpers/html/config.php:
```diff
@@ -0,0 +1,55 @@
+	/**
+	 * Display a clickable list of tags for a mail template
+	 *
+	 * @param   string  $mail       Identifier of the mail template.
```
this actually a `CMSObject`

Fedik commented on administrator/components/com_config/helpers/html/config.php:
```diff
@@ -0,0 +1,55 @@
+	public static function mailtags($mail, $fieldname)
+	{
+		$app = Factory::getApplication();
+		Factory::getApplication()->triggerEvent('onMailBeforeTagsRendering', array($mail->mail_id, &$mail));
```
since `$mail` is `CMSObject` you do not required to use reference `&$mail`, enough `$mail`

Fedik commented:
So I have tested (Text version) and it looks good to me (except broken editing of HTML version).
I was able to add custom Tags via plugin events, and all worked:

```php
/**
 * @param string    $mail_id
 * @param CMSObject $mail
 */
public function onMailBeforeTagsRendering($mail_id, $mail)
{
  if ($mail_id === 'com_config.test_mail')
  {
    $mail->params['tags']['beer'] = 'beer';
    $mail->params['tags']['bear'] = 'bear';
  }
}

/**
 * @param string       $mail_id
 * @param MailTemplate $mailTemplate
 */
public function onMailBeforeRendering($mail_id, $mailTemplate)
{
  if ($mail_id === 'com_config.test_mail')
  {
    $mailTemplate->addTemplateData(array(
      'beer' => 'Beeeer',
      'bear' => 'Beeeeaaaaaar'
    ));
  }
}
```

Couple notes:
- How the extension can add its own template? I think that forcing them to use SQL is not gut idea. 
- Would be cool to have "send test", while editing the email

Fedik commented on libraries/src/Mail/MailTemplate.php:
```diff
@@ -0,0 +1,219 @@
+		$db->setQuery($query);
+		$mail = $db->loadObject();
+
+		Factory::getApplication()->triggerEvent('onMailBeforeRendering', array($this->mail_id, &$this));
```
and here same, you do not need `&$this`, just `$this`, 
it is object, they all by reference :wink:

Fedik commented:
Fix for Multiple editors #22341

Hackwar committed:
Merge branch '4.0-dev' into j4mailtemplates

wilsonge committed:
Code cleanup

Hackwar committed:
Merge branch '4.0-dev' into j4mailtemplates

Hackwar committed:
Merge branch '4.0-dev' into j4mailtemplates

laoneo committed:
Merge branch '4.0-dev' into j4mailtemplates

laoneo commented:
What about the possibility to offer components to add the mailer list to their submenu? People install an extension and then they want to set it up. Editing mail templates is one task of it. Like it is now you have to go to a menu item buried somewhere in the system menu. Similar to what we have now with categories and fields. The user can stay in the context.

laoneo commented:
Beside that I really love that functionality. Code review will come later, except that I would move the it to com_mailto.

laoneo commented:
@Hackwar did you see my comments? Would love to proceed with this one.

Hackwar commented:
@laoneo I'm torn on your proposal. I understand your proposal, but I do have two issues with going that way. The first is, that there are components that only send one mail (com_config, com_joomlaupdate) and it is pretty much overkill to have an extra submenu entry, list view, etc. for just this one mail template and the second is that this would only work for components. But I want this to be also used by plugins. For example, a customer of mine uses a plugin to send a mail to a user when the user is deleted. I don't know how that one would be accessible via the GUI for a plugin.

Maybe we can simply keep it the way it is now and add a second mode, where it filters for a certain extension and displays that extensions submenu when called by a specific parameter? That could be implemented in a second step when this one is merged.

Fedik commented:
> What about the possibility to offer components to add the mailer list to their submenu?

I am agree with @Hackwar , this not really need for now, just a filter can be enough.
And I think it will be not that often used thing, like tags or fields. 
Just an one more configuration possibilities.

laoneo commented:
So at least filters should be added before merging.

laoneo committed:
Merge branch '4.0-dev' into j4mailtemplates

Hackwar committed:
Merge branch '4.0-dev' into j4mailtemplates

Hackwar committed:
Adding filters

Hackwar committed:
Codestyle

Hackwar commented:
Filters have been added. Please consider this for merging and then we can add a bunch of other templates to this and add more features...

Hackwar committed:
Merge branch '4.0-dev' into j4mailtemplates

Hackwar committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into j4mailtemplates

# Conflicts:
#	administrator/components/com_cpanel/View/System/HtmlView.php

Hackwar committed:
Merge branch 'j4mailtemplates' of github.com:Hackwar/joomla-cms into j4mailtemplates

Hackwar committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into j4mailtemplates

Hackwar committed:
Adding mail templating system

Hackwar commented:
I've basically started over with this PR and did a lot of changes. There are still a few JS issues in the edit screen, where I would kindly ask for some help. I'm just not really a JS guy. Please read the initial text of this PR to know about all the changes that I did.

Fedik commented:
@Hackwar do as you can, then write what issue you have with JS, and I will try to help when will get some time

Fedik commented on administrator/language/en-GB/en-GB.com_mails.ini:
```diff
@@ -0,0 +1,56 @@
+COM_MAILS_CONFIG_FIELD_ALTERNATIVE_MAILCONFIG_LABEL="Allow alternative mail configuration"
+COM_MAILS_CONFIG_FIELD_COPY_MAILS_LABEL="Send copy of mails"
+COM_MAILS_CONFIG_FIELD_COPY_MAILS_LABEL
+COM_MAILS_CONFIG_FIELD_ALTERNATIVE_MAILCONFIG_LABEL
```
these two looks empty :wink:

Hackwar replied:
Oops. I will fix that soon.

Hackwar commented:
@Fedik The switch to enable/disable the input fields when editing a mail template don't work right and you can click on a tag and it is added to the input field, even though they are disabled.

I have no idea how we can add the nested tags in a good way into the UI. Maybe that is really something for a future PR... I also don't know how we can handle the same tags in plaintext and html. I'm open for suggestions (and contributions to this PR)

Hackwar committed:
Codestyle and language fix

joomdonation commented:
I had a quick look at the PR and have few feedback/questions:

1. Look like we removed two fields custom_data, system_data from #__extensions table, so the installation SQL in this PR won't work. You need to remove two fields from INSERT command

2. The code in **edit**, **save** methods of **TemplateController** and **getItem**, **save** methods of TemplateModel is almost the same/duplicate code of it's parent class. Why is that? I guess it's because you don't use id as primary key of #__mail_templates table instead of using id as most of Joomla tables. Personal, I don't like these duplicate code, so unless there are special reason, I would use id as primary key to remove these duplicate code.

3. We would have to find out how to handle multilingual website. From what I see, for each language, we need one record for a mail template. So how we populate these records on install? And how when user installs a new language to the site?

4. I have the feeling that the MailTemplate is doing more than it needs. I could not think of a case when I could define attachments, BCC/CC emails... when the mail template is inserted into database. If I have to add attachments, I would do that in $mail object before passing it to MailTemplate, same for BCC/CC emails. These data usually dynamic data, not static.

These are few minor issues with the code, but I will review it later, depends on the decision for #2

Hackwar commented:
to 1. I will have a look at that.

to 2. That is related to using a master template and then the different languages. The key for a template is always the identifier `com_config.test_mail` and then the language. I don't like that duplicate code either, but it is far better than the alternative of introducing an unnecessary integer ID here.

to 3. This IS ready for multilingual websites. Each mail template in the beginning consists of a master template that has translateable language strings. If there is no language specific mail template, then it will use the master template and simply translate the language strings. You can then go and edit the master template for a specific language and save that and it will generate a language specific template. That is why there are switches next to the subject and bodys. You can switch using those language strings or your custom template on or off.

to 4. The static attachments is aimed at shops, etc. which want to send out their ToS as a PDF via mail. I could come up with several more scenarios. The BCC I personally would use to have a log of the mails that my system has sent out and to be able to test this over a certain period of time.

parthlawate commented:
Hello Sorry I'm jumping on this a bit late. There is a PR we are currently working on to merge our tj-notifications system in Joomla to allow easy editing of all system emails and provide a centralised notification system. My Original detailed blog on this is lost as it was hosted on joomla.com and i missed renewing it :( . However the GitHub can be checked here : https://github.com/techjoomla/TJ-Notifications/wiki This is in use now for 2 odd years in production and is pretty stable.. 

There is a little bit of an overlap on this PR on it. In this PR we would simply need to add the Notification magic method in all places that currently triggers an email. This will allow further extension to notifications via SMS, Push etc. 

In addition to the notifications, I would also be willing to donate the code of Email Beautifier (or merge it in this PR to bring any features we have there that are missing here). 

cc @manojLondhe @techjoomla @coolbung

uglyeoin commented:
I tried to use patch tester to install this and I got this message ×
Error
The file marked for modification does not exist: **administrator/components/com_config/Model/ApplicationModel.php**

Hackwar commented:
Are you using this on a Joomla 4.0 installation?

uglyeoin commented:
> Are you using this on a Joomla 4.0 installation?

Ah of course, whoops

Fedik committed:
EmailTemplateEdit js

Fedik committed:
Show/hide tags section. Do not allow to add tag in to disabled input

Fedik commented:
@Hackwar please look https://github.com/Hackwar/joomla-cms/pull/26

A note: I would remove `MailsHelper::mailtags` (it hardcoded HTML) and use a "sub template" instead (eg `edit_mailtaglist.php`).

One issue still left, with HTML Body editor. If it initialized as Disabled it not possible to enable it, we do not have an API for it.
Maybe we can leave it always enabled (on initial state) for now or?

Hackwar committed:
Merge pull request #26 from Fedik/pull22126

 Email templates. Show/hide tags section. Do not allow to add tag in to disabled input

Hackwar committed:
Merge branch '4.0-dev' into j4mailtemplates

Hackwar committed:
Merge branch '4.0-dev' into j4mailtemplates

sanek4life commented:
please, add my AMP-mail feature request #24393

Hackwar commented:
Your request has nothing to do with this PR and is completely unrelated. You also already got a decision in your issue regarding that.

Hackwar committed:
Merge branch '4.0-dev' into j4mailtemplates

Hackwar committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into j4mailtemplates

# Conflicts:
#	administrator/language/en-GB/en-GB.com_config.ini

Hackwar committed:
Merge branch 'j4mailtemplates' of github.com:Hackwar/joomla-cms into j4mailtemplates

# Conflicts:
#	administrator/components/com_cpanel/Model/SystemModel.php
#	administrator/language/en-GB/en-GB.com_config.ini
#	installation/sql/mysql/joomla.sql
#	installation/sql/postgresql/joomla.sql
#	libraries/src/Mail/MailTemplate.php

Hackwar committed:
JS Codestyle

Hackwar committed:
JS Codestyle

Hackwar committed:
Fixing SQL install files

Hackwar committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into j4mailtemplates

Hackwar committed:
Fixing SQL

wilsonge commented on administrator/components/com_config/forms/mail.xml:
```diff
@@ -0,0 +1,72 @@
+		/>
+
+		<field
+			name="subject_switcher"
```
Seems to be missing a label in the UI :)

wilsonge replied:
<img width="1278" alt="Screenshot 2019-06-26 at 00 22 19" src="https://user-images.githubusercontent.com/1986000/60140146-8731d280-97a8-11e9-8ca8-af2e1e9e3758.png">

wilsonge commented on administrator/components/com_config/View/Mail/HtmlView.php:
```diff
@@ -0,0 +1,130 @@
+	/**
+	 * The active item
+	 *
+	 * @var  object
```
state and item both CMSObjects

uglyeoin commented:
I installed via patch tester and then I also discovered com_mail but when I go to the system link it says 404 component not found.  Did I do something incorrectly?  Also the language string under system is broken. I don't find anything in components only in system, but I think that is right according to the PR?

Hackwar committed:
Merge branch '4.0-dev' into j4mailtemplates

Hackwar committed:
Merge branch '4.0-dev' into j4mailtemplates

Hackwar committed:
Codestyle and cleanup of com_mails

wilsonge committed:
Merge branch '4.0-dev' into j4mailtemplates

Hackwar committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into j4mailtemplates

Hackwar committed:
Fixing merge conflicts

Hackwar committed:
Merge branch 'j4mailtemplates' of github.com:Hackwar/joomla-cms into j4mailtemplates

wilsonge committed:
Fix javascript to remove switcher references

wilsonge committed:
Remove console log

Hackwar committed:
Fix javascript to remove switcher references (#30)

Fix javascript to remove switcher references

Hackwar committed:
Fixing readonly for textarea
Hiding options tab when no options available

Hackwar commented:
This PR has unfortunately gathered some dust in the last few months... Today @wilsonge and I tried to make it all work again and we partially succeeded. The component works 98%, but there are 2 issues remaining that I can't fixed:

- The switcher elements have a legend attached to them, which I'd rather have not visible. How can I remove that? @brianteeman?
- The WYSIWYG editors (or specifically TinyMCE) don't react on the change to the underlying textarea element. So switching the field on or off does not work. I don't know how we can best implement this. I would actually expect that a change to the readonly attribute of a textarea results in the editor making this editable again.

If someone could help out with this, I'd be really happy.

brianteeman commented on administrator/components/com_config/Model/ApplicationModel.php:
```diff
@@ -1027,10 +1029,16 @@ public function sendTestMail()
 		$mail = Factory::getMailer();
 
 		// Prepare email and send try to send it
```
```suggestion
		// Prepare email and try to send it
```

brianteeman commented on administrator/components/com_mails/Controller/DisplayController.php:
```diff
@@ -0,0 +1,30 @@
+/**
+ * Controller for global configuration
+ *
+ * @since  1.5
```
shouldnt this be _DEPLOY_VERSION_ or 4.0 ?

brianteeman commented on administrator/components/com_mails/Helper/MailsHelper.php:
```diff
@@ -0,0 +1,57 @@
+<?php
+/**
+ * @package     Joomla.Administrator
+ * @subpackage  com_config
```
```suggestion
 * @subpackage  com_mails
```

brianteeman commented on administrator/components/com_mails/config.xml:
```diff
@@ -0,0 +1,66 @@
+
+		<field
+			name="attachment_folder"
+			type="text" 
```
shouldnt this use type="folderlist" ?

Hackwar replied:
In theory yes, in reality I don't want to force people to use a specific folder. So I would have to use the root of the Joomla installation. That list of folders and subfolders would be freakishly long. Thus I chose to use a text field. If there is a better alternative, I'm all ears.

brianteeman commented on administrator/components/com_mails/forms/filter_templates.xml:
```diff
@@ -0,0 +1,55 @@
+		<field
+			name="language"
+			type="contentlanguage"
+			onchange="this.form.submit();"
```
```suggestion
			label="JOPTION_SELECT_LANGUAGE"
			onchange="this.form.submit();"
```

brianteeman commented on administrator/components/com_mails/forms/filter_templates.xml:
```diff
@@ -0,0 +1,55 @@
+		<field
+			name="extension"
+			type="sql"
+			query="SELECT SUBSTRING(template_id, 1, POSITION('.' IN template_id) - 1) AS value, SUBSTRING(template_id, 1, POSITION('.' IN template_id) - 1) AS extension FROM #__mail_templates GROUP BY extension ORDER BY extension ASC"
```
please add a label

brianteeman commented on administrator/components/com_mails/forms/filter_templates.xml:
```diff
@@ -0,0 +1,55 @@
+		<field
+			name="fullordering"
+			type="list"
+			onchange="this.form.submit();"
```
```suggestion
			label="JGLOBAL_SORT_BY"
			onchange="this.form.submit();"
```

brianteeman commented on administrator/components/com_mails/forms/filter_templates.xml:
```diff
@@ -0,0 +1,55 @@
+		<field
+			name="limit"
+			type="limitbox"
+			default="25"
```
```suggestion
			label="JGLOBAL_LIST_LIMIT"
			default="25"
```

brianteeman commented on administrator/components/com_mails/forms/template.xml:
```diff
@@ -0,0 +1,247 @@
+			name="subject"
+			type="text" 
+			label="COM_MAILS_FIELD_SUBJECT_LABEL"
+			class="input-xxlarge"
```
this line can be removed as its not used in j4
```suggestion
```

brianteeman commented on administrator/components/com_mails/forms/template.xml:
```diff
@@ -0,0 +1,247 @@
+		<field
+			name="subject_switcher"
+			type="radio"
+			class="switcher"
```
please add a label

brianteeman commented on administrator/components/com_mails/forms/template.xml:
```diff
@@ -0,0 +1,247 @@
+		<field
+			name="body_switcher"
+			type="radio"
+			class="switcher"
```
please add a label

brianteeman commented on administrator/components/com_mails/forms/template.xml:
```diff
@@ -0,0 +1,247 @@
+		<field
+			name="htmlbody_switcher"
+			type="radio"
+			class="switcher"
```
please add a label

brianteeman commented on administrator/components/com_mails/forms/template.xml:
```diff
@@ -0,0 +1,247 @@
+		label="COM_MAILS_FIELD_ATTACHMENTS_LABEL"
+		multiple="true"
+		layout="joomla.form.field.subform.repeatable-table"
+	>
```
```suggestion
	/>
```

brianteeman commented on administrator/components/com_mails/forms/template.xml:
```diff
@@ -0,0 +1,247 @@
+				name="smtpauth"
+				type="radio"
+				class="switcher"
+				label="COM_MAILS_FIELD_MAIL_SMTP_AUTH_LABEL"
```
swap with l207

brianteeman commented on administrator/components/com_mails/services/provider.php:
```diff
@@ -0,0 +1,54 @@
+<?php
+/**
+ * @package     Joomla.Administrator
+ * @subpackage  com_config
```
```suggestion
 * @subpackage  com_mails
```

brianteeman commented:
> The switcher elements have a legend attached to them, which I'd rather have not visible. How can I remove that? @brianteeman?

@hackwar are you referring to this
![image](https://user-images.githubusercontent.com/1296369/62010054-c368a200-b15d-11e9-93a4-ebf24c0217f4.png)

Thats because you dont have a label in the xml.

brianteeman commented:
![image](https://user-images.githubusercontent.com/1296369/62010094-193d4a00-b15e-11e9-828c-df891d77dae1.png)

Please don't do this. At a minimum it should be just the text. If you **want** to add a flag as well then it must obey the setting in content_languages

Hackwar commented:
> 
> 
> > The switcher elements have a legend attached to them, which I'd rather have not visible. How can I remove that? @brianteeman?
> 
> @Hackwar are you referring to this
> ![image](https://user-images.githubusercontent.com/1296369/62010054-c368a200-b15d-11e9-93a4-ebf24c0217f4.png)
> 
> Thats because you dont have a label in the xml.

When I'm adding a label, it is still displaying that label. I don't want to have anything there, just the switcher.

brianteeman commented:
How would anyone know what the switch is for?

infograf768 commented on administrator/language/en-GB/en-GB.com_mails.ini:
```diff
@@ -0,0 +1,55 @@
+COM_MAILS_CONFIG_MAIL_OPTIONS="Mails"
+COM_MAILS_CONFIG_FIELD_ALTERNATIVE_MAILCONFIG_LABEL="Allow alternative mail configuration"
+COM_MAILS_CONFIG_FIELD_COPY_MAILS_LABEL="Send copy of mails"
+COM_MAILS_CONFIG_FIELD_ALTERNATIVE_MAILCONFIG_LABEL="Alternative Mail-configuration"
```
This constant exists twice.

infograf768 commented:
Not sure I understand all of this, as it looks pretty hard for non-coders to create their own mail template, but here is a remark concerning language based on the example proposed for send mail test.

Using `type="contentlanguage"` looks wrong to me because we may have a site with multiple languages but NOT set as a multilingual site, i.e. where the Content Languages have been deleted because they are useless (thereefore no specific flag or sef).

In the case of sending test mail, which is done from backend, it should imho just depend on the admin language in use. Therefore filtering by Content Language as well as using   `getContentLanguages` instead of `getInstalledLanguages` with client administrator is wrong.

Hackwar commented:
There is no interest in letting normal users create their own templates. This is a way for an extension to allow for complex mails with translated content. You can edit existing templates, not create new ones.

We don't know if a template is used in the frontend or backend when editing the template. So I wont know if to use ```isAdminMultilang```. The same issue is with the contentlanguages vs installed languages. I have to choose one of the two for the form XML and as we had this discussion in the past, I chose the content languages, among other reasons because you might have several content languages for one installed language.

infograf768 commented:
> I chose the content languages, among other reasons because you might have several content languages for one installed language.

It is wrong in this test case as I clearly demonstrated.

Hackwar committed:
Several fixes

wilsonge commented on administrator/components/com_mails/forms/template.xml:
```diff
@@ -61,8 +60,10 @@
 		<field
 			name="htmlbody_switcher"
 			type="radio"
+			label="sdfaghg"
```
:)

Hackwar committed:
Removing accidental label

Hackwar committed:
Adding labels, fixing docblocks

Hackwar committed:
Merge branch 'j4mailtemplates' of github.com:Hackwar/joomla-cms into j4mailtemplates

# Conflicts:
#	administrator/components/com_mails/forms/template.xml

brianteeman commented:
Now that you have added labels I am even more confused about what they do

Hackwar commented:
> 
> 
> Now that you have added labels I am even more confused about what they do

How about you click them?

brianteeman commented:
>> Now that you have added labels I am even more confused about what they do

> How about you click them?

You shouldnt have to in order to understand their purpose. But I will install the branch and look

Fedik commented:
maybe something like "Customize HTML body" or "Edit HTML email" can be better for label?

Hackwar commented:
I think the best solution would be to write a ten page essay as label...

brianteeman commented:
Might have helped me if I had noticed that there was some JS in this PR - oops

Edit Subject yes/no
Edit Body yes/no

Fedik commented:
@brianteeman email has 2 body, for a plain text and a html.
"Edit Body yes/no" maybe not enough

brianteeman commented:
Edit Subject yes/no
Edit Body yes/no
Edit HTML Body yes/no

Hackwar committed:
Clarifying language strings used

Hackwar committed:
Merge branch '4.0-dev' into j4mailtemplates

Hackwar commented:
done

brianteeman commented:
Thanks

alikon commented:
is this a feature to be added in core ? 
shouldn't it be matter for 3dp extension ?

Fedik commented:
> is this a feature to be added in core ?

I would like to see it in the core,
Joomla!  still send all its emails in plain text (while other CMS use nicer HTML emails)
and this feature gives an opportunity to be able to style it somehow nicer.

infograf768 commented:
I do like the feature, but not the way Languages are treated. 
It should not depend on Content Languages which are/should be used only used for Multilingual sites. 
It should only depend on Installed Languages (site or admin depending on where the email is sent from)

Fedik commented:
> It should not depend on Content Languages

I can understand a reason why it made like that.

@Hackwar @infograf768  what do you think about next approach:
- by default the template sets to all languages, `language = *`
- inside the template allow to use placeholders like {TEXT:FOO_BAR_LANG_STRING} for translatable strings, eg:
```
<h1>{TEXT:HELLO_USER}</h1>
<p>{TEXT:WELCOME_MESSAGE}</p>
```

In theory it will allow to simplify the thing for Users a bit, will use a default (active) language, and will separate Site/Admin.

Is it will be hard to implement?

Hackwar commented:
I wonder if you guys actually looked at this in depth. You don't have to create any templates at all, because it will always fall back on a translation string. **IF** you create your own template, it takes that translation string, translates it and prepopulates the respective input with that. In that translated string, you already have placeholders for the actual data. You can then reorder that whichever way you want or add it multiple times. It even contains the code to use subtemplates, even though that is not available via the GUI yet. Something for 4.1 or 4.2.

I don't see any reason to change it to what @Fedik described above.

brianteeman commented:
The one thing that is missing to make this truly useful is the ability to send yourself a test email

wilsonge commented:
> The one thing that is missing to make this truly useful is the ability to send yourself a test email

Seems slightly non-trivial because you'd need to manually add all the data that can be injected with tags into the body of the email.

Hackwar committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into j4mailtemplates

# Conflicts:
#	administrator/components/com_cpanel/Model/SystemModel.php
#	administrator/language/en-GB/en-GB.mod_menu.ini

roland-d commented:
@Hackwar Is this ready for testing? If so, could you please resolve the conflict?

wilsonge committed:
Merge branch '4.0-dev' into j4mailtemplates

wilsonge commented:
Conflict fixed. As far as I'm concerned this will do for feature completeness for now. I kind of agree with @brianteeman 's comment about test mails but also understand that it can be complicated for this and am happy to leave this for a future Joomla version.

I believe the only thing left here beyond testing is to make a final call whether to use content languages or installed languages for the emails (I leave that to people who actually build multilang sites) - probably one for a Production Department call @HLeithner ?

roland-d commented:
@wilsonge For me, a test email doesn't need actual values, in general you just want to see how the mail will look like.

wilsonge commented:
You still need some sort of value however which is going to be context specific

Hackwar commented:
@wilsonge Did you fix the issue with the HTML editor enabling/disabling?

alikon commented on administrator/components/com_admin/sql/updates/postgresql/4.0.0-2019-01-16.sql:
```diff
@@ -0,0 +1,17 @@
+INSERT INTO "#__extensions" ("name", "type", "element", "folder", "client_id", "enabled", "access", "protected", "manifest_cache", "params", "custom_data", "system_data", "checked_out", "checked_out_time", "ordering", "state") VALUES
+('com_mails', 'component', 'com_mails', '', 1, 1, 1, 1, '{"name":"com_mails","type":"component","creationDate":"January 2019","author":"Joomla! Project","copyright":"(C) 2005 - 2018 Open Source Matters. All rights reserved.","authorEmail":"admin@joomla.org","authorUrl":"www.joomla.org","version":"4.0.0","description":"COM_MAILS_XML_DESCRIPTION","group":""}', '{}', '', '', 0, '0000-00-00 00:00:00' ...
```
```suggestion
('com_mails', 'component', 'com_mails', '', 1, 1, 1, 1, '{"name":"com_mails","type":"component","creationDate":"January 2019","author":"Joomla! Project","copyright":"(C) 2005 - 2019 Open Source Matters. All rights reserved.","authorEmail":"admin@joomla.org","authorUrl":"www.joomla.org","version":"4.0.0","description":"COM_MAILS_XML_DESCRIPTION","group":""}', '{}', '', '', 0, '1970-01-01 00:00:00', 0, 0);
```

Hackwar committed:
Update administrator/components/com_admin/sql/updates/postgresql/4.0.0-2019-01-16.sql

Co-Authored-By: Nicola Galgano <optimus4joomla@gmail.com>

HLeithner commented:
> I believe the only thing left here beyond testing is to make a final call whether to use content languages or installed languages for the emails (I leave that to people who actually build multilang sites) - probably one for a Production Department call @HLeithner ?

My opinion is that it should be content specific but last time I tried (it's a long time ago) content languages with out a similar installed language doesn't work properly so I didn't used it.

bees4ever commented:
Having here the "Pizza-Bugs-and-Fun" with @roland-d I tried to get this PR tested. However, I had a clean installation of Joomla 4 and applied the patch. First, the extension didn't get corrected installed, i.e. the mysql DB was not updated correctly, when I opened http://localhost/joomla-cms3/administrator/index.php?option=com_mails&view=templates locally this error appeard.

<img width="480" alt="Bildschirmfoto 2019-09-02 um 15 49 24" src="https://user-images.githubusercontent.com/21102167/64119405-14da0180-cd9a-11e9-8dec-935510500d97.png">


I updated the DB myself but then this error appeared:

<img width="863" alt="Bildschirmfoto 2019-09-02 um 15 52 10" src="https://user-images.githubusercontent.com/21102167/64119412-17d4f200-cd9a-11e9-9130-d5dd8d2e05fa.png">


What can I do to get the patch run?

Thanks

roland-d committed:
Remove obsolete columns from SQL statements
Minor code cleanup

Signed-off-by: Roland Dalmulder <contact@rolandd.com>

roland-d commented:
@Hackwar I send PR https://github.com/Hackwar/joomla-cms/pull/33 to fix the SQL errors.

Hackwar committed:
Merge pull request #33 from roland-d/code-fixes-j4

Remove obsolete columns from SQL statements

Hackwar commented:
thx @roland-d

Schmidie64 commented:
You should add to the test instructions that you have to delete the autoload_psr4.php file (/ libraries) and the Javascript files have to be rebuilt (npm run build: js)

wilsonge commented on administrator/components/com_mails/services/provider.php:
```diff
@@ -41,7 +41,7 @@ public function register(Container $container)
 		$container->set(
 			ComponentInterface::class,
-			function (Container $container)
+			static function (Container $container)
```
This shouldn't be static - it's an anonymous function

Hackwar committed:
Removing false static

Hackwar committed:
Merge branch '4.0-dev' into j4mailtemplates

wilsonge committed:
Fix tags not showing in html mode

wilsonge committed:
Cleanup template model doc block

wilsonge committed:
Start work on disabled function

- Add concept to tinymce - requires addition of initalising in readonly
- Add concept finshed to editor none
- TODO: Codemirror

wilsonge committed:
If textarea has readonly attribute start tinymce in readonly mode

wilsonge committed:
Add function to codemirro

wilsonge committed:
Simplify editor none

Hackwar committed:
Editor disabled (#35)

Editor disabled

Hackwar committed:
Merge branch '4.0-dev' into j4mailtemplates

wilsonge committed:
Merge branch '4.0-dev' into j4mailtemplates

wilsonge commented:
Merging this as a starting point as the basics work. I fully accept there needs to be some work here as we move through getting 4.0 to stable. Personally for example the way if you enable html mails don't show the text at all for the sample mail is confusing (although having dived through the code I think I finally understand the logic of why). So let's start with this get some user testing done as it ends up in the nightlies and see how things go.

micker commented:
hello thanks for this great feature a question, no plan to use with new workflow system ?
link transition action to a specific email template will be a great function
both feature will be improve and polished
thanks @Hackwar

Hackwar commented:
No, right now the workflow is not implemented here. I'm open to improve this in the future; for 4.0 I would keep it the way it is and postpone this to 4.1 or similar.

micker commented:
ok cool ! thanks i will be a great function !!!
